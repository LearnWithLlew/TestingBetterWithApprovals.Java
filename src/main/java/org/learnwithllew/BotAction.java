package org.learnwithllew;

import java.util.List;

public record BotAction(List<Command> commands) {

}
