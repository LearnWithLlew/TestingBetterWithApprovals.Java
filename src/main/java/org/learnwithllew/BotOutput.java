package org.learnwithllew;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BotOutput {
    private final Map<String, BotAction> messages = new HashMap<>();

    public void send(String conversationId, BotAction action) {
        messages.put(conversationId, action);
    }

    public BotAction read(String conversationId) {
        return messages.get(conversationId);
    }
}

