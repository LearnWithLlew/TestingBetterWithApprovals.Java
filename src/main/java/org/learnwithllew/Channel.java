package org.learnwithllew;

public enum Channel {
    TEXT,
    VOICE
}
