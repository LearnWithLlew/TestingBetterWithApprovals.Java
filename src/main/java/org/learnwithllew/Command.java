package org.learnwithllew;

public interface Command {
}
