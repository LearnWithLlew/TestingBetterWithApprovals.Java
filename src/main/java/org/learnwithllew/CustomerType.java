package org.learnwithllew;

public enum CustomerType {
    UNKNOWN,
    EXISTING,
    PROSPECTIVE
}
