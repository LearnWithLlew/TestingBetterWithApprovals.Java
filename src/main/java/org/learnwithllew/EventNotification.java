package org.learnwithllew;

import org.lambda.query.Queryable;

import java.util.List;

public class EventNotification {
    private final String conversationId;
    private final UserId userId;
    private final List<Property> properties;
    private final Queryable<MessageEvent> events;

    private EventNotification(UserId userId, List<Property> properties, Queryable<MessageEvent> events, String conversationId) {
        this.userId = userId;
        this.properties = properties;
        this.events = events;
        this.conversationId = conversationId;
    }

    public UserId getUserId() {
        return userId;
    }

    public List<Property> getProperties() {
        return properties;
    }

    public List<MessageEvent> getEvents() {
        return events;
    }

    public String getConversationId() {
        return conversationId;
    }

    public static class Builder {
        private String conversationId;
        private UserId userId;
        private List<Property> properties;
        private Queryable<MessageEvent> events;

        public Builder conversationId(String conversationId) {
            this.conversationId = conversationId;
            return this;
        }

        public Builder userId(UserId userId) {
            this.userId = userId;
            return this;
        }

        public Builder properties(List<Property> properties) {
            this.properties = properties;
            return this;
        }

        public Builder events(List<MessageEvent> events) {
            this.events = Queryable.as(events);
            return this;
        }

        public EventNotification build() {
            return new EventNotification(userId, properties, events, conversationId);
        }
    }
}

