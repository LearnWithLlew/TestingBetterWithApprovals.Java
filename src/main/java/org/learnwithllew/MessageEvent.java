package org.learnwithllew;

public record MessageEvent(long time, String id, String message) {
}
