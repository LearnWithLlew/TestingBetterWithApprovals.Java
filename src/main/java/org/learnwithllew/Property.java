package org.learnwithllew;

public record Property(String name, String value) {
}
