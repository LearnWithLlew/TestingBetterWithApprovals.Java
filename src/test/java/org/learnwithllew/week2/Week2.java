package org.learnwithllew.week2;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.learnwithllew.Bot;
import org.learnwithllew.BotOutput;

// Full printed story -> whitespace -> json
public class Week2 {

    @Test
    void botClarifiesIntentOnGreeting() {
        verifyConversation("hi");
    }

    @Test
    void repeatedGreetingGetsAnotherPrompt() {
        verifyConversation("hi", "hi");
    }

    @Test
    void botOffersSelfServiceOnSupportedIntents() {
        verifyConversation("hi", "pay bill", "Yes, I'm a customer");
    }

    private void verifyConversation(String... messages) {
        BotOutput output = new BotOutput();
        Bot bot = new Bot(output);
        var storyBoard = StoryBoard.create(bot, output, messages);
        Approvals.verify(storyBoard);
    }
}
