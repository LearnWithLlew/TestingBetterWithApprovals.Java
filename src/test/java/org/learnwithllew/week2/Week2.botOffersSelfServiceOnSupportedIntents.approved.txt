[Customer]: hi
[Bot]: Hi there! I'm your virtual assistant.
[Bot]: What would you like to do today?
[Customer]: pay bill
[Bot]: Let me try to help you.
[Bot]: {
  "content": {
    "type": "vertical",
    "elements": [
      {
        "type": "text",
        "text": "Are you a customer?",
        "tooltip": "text tooltip"
      },
      {
        "type": "horizontal",
        "elements": [
          {
            "type": "button",
            "title": "Yes",
            "click": {
              "actions": [
                {
                  "type": "publishText",
                  "text": "Yes, I'm a customer"
                }
              ]
            }
          },
          {
            "type": "button",
            "title": "No",
            "click": {
              "actions": [
                {
                  "type": "publishText",
                  "text": "No, I'm not"
                }
              ]
            }
          }
        ]
      }
    ]
  }
}
[Customer]: Yes, I'm a customer
[Bot]: transfers to 'self_service'
